src/item.h

```diff
@@ -1194,25 +1194,47 @@ class item : public JsonSerializer, public JsonDeserializer, public visitable<it
         item * gunmod_find( const itype_id& mod );
         const item * gunmod_find( const itype_id& mod ) const;
 
-        /**
-         * Returns currently active auxiliary (@ref is_auxiliary_gunmod) gun mod item.
-         * May return null if there is no such gun mod or if the gun is not in the
-         * auxiliary mode (@ref is_in_auxiliary_mode).
-         */
-        item * gunmod_current();
-        item const * gunmod_current() const;
-
         /*
          * Checks if mod can be applied to this item considering any current state (jammed, loaded etc.)
          * @param alert whether to display message describing reason for any incompatibility
          * @param effects whether temporary efects (jammed, loaded etc) are considered when checking
          */
         bool gunmod_compatible( const item& mod, bool alert = true, bool effects = true ) const;
 
-        /**
-         * Burst size (see ranged.cpp), includes effects from installed gunmods.
-         */
-        int burst_size() const;
+        struct gun_mode {
+            std::string mode; /** name of this mode */
+            item *target;     /** pointer to base gun or attached gunmod */
+            int qty;          /** burst size or melee range */
+            bool melee;       /** if true perform a melee attach as opposed to shooting */
+
+            operator bool() const { return target != nullptr; }
+
+            item &operator*() { return *target; }
+            const item &operator*() const { return *target; }
+
+            item *operator->() { return target; }
+            const item *operator->() const { return target; }
+        };
+
+        /** Get all possible modes for this gun inclusive of any attached gunmods */
+        std::map<std::string, const item::gun_mode> gun_all_modes() const;
+
+        /** Check if gun supports a specific mode returning an invalid/empty mode if not */
+        const gun_mode gun_has_mode( const std::string& mode ) const;
```

Comment:
> This function tests for an existence of a mode so has_ is appropriate.

This goes against 99% of our code:
* `has_quality` doesn't return the item with said quality
* `has_item_with` doesn't return the item
* `has_effect` doesn't return the effect
* `has_special_attack` doesn't return the special attack
* `map::has_flag` doesn't return the terrain/furniture that sets the flag
* `vehicle::has_structural_part` doesn't return the index or reference of the part

This is non-standard.

Additionally, typical usage of the function is:
```
gun_has_mode( "AUTO" ).qty
```
Which clearly goes against the `has_` in name.

At least rename it to `gun_get_mode`, so that it isn't confusing and against the (non-written, but still) standards.